server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -680,72 +689,156 @@ public String toString() {
+        private final LocalCheckpointTracker requestSeqIdTracker = new LocalCheckpointTracker(NO_OPS_PERFORMED, NO_OPS_PERFORMED);
+        private final Deque<byte[]> recycledBuffers;
+        private boolean readerClosed; // ensure we don't reopen reader after the recovery was cancelled
+        private StoreFileMetaData md;
+        private InputStream currentInput;
+        private int position;
+        private final Object turnMutex = new Object();
+        private final AtomicBoolean turnAcquired = new AtomicBoolean();
+
+        MultiFileSender(Store store, StoreFileMetaData[] files, Supplier<Integer> translogOps) {
+            this.store = store;
+            this.files = new ArrayList<>(Arrays.asList(files));
+            this.translogOps = translogOps;
+            this.recycledBuffers = ConcurrentCollections.newDeque();
+            for (int i = 0; i < maxConcurrentFileChunks; i++) {
+                this.recycledBuffers.add(new byte[chunkSizeInBytes]);
             }
-            try (IndexInput indexInput = store.directory().openInput(md.name(), IOContext.READONCE);
-                 InputStream in = new InputStreamIndexInput(indexInput, md.length())) {
-                long position = 0;
-                int bytesRead;
-                while ((bytesRead = in.read(buffer, 0, buffer.length)) != -1) {
-                    final BytesArray content = new BytesArray(buffer, 0, bytesRead);
-                    final boolean lastChunk = position + content.length() == md.length();
-                    final long requestSeqId = requestSeqIdTracker.generateSeqNo();
-                    cancellableThreads.execute(() -> requestSeqIdTracker.waitForOpsToComplete(requestSeqId - maxConcurrentFileChunks));
-                    cancellableThreads.checkForCancel();
-                    if (error.get() != null) {
-                        break;
+        }
+
+        void sendFileChunks(ActionListener<Void> listener) {
+            while (true) {
+                synchronized (turnMutex) {
+                    if (turnAcquired.compareAndSet(false, true) == false) {
+                        return;
                     }
-                    final long requestFilePosition = position;
-                    cancellableThreads.executeIO(() ->
-                        recoveryTarget.writeFileChunk(md, requestFilePosition, content, lastChunk, translogOps.get(),
+                    if (requestSeqIdTracker.getMaxSeqNo() - requestSeqIdTracker.getCheckpoint() >= maxConcurrentFileChunks) {
+                        turnAcquired.set(false);
+                        return;
+                    }
+                }
+                try {
+                    final byte[] buffer = recycledBuffers.removeFirst();
+                    final FileChunk chunk = readChunk(buffer);
+                    if (chunk == null) {
+                        if (requestSeqIdTracker.getMaxSeqNo() == requestSeqIdTracker.getCheckpoint()) {
```

Comment:
I don't get this eihter why is the seqIdTracker invovlded here we only need to send files no?